pennylane/gradients/jvp.py

```diff
@@ -0,0 +1,287 @@
+    vector of gradient outputs and a Jacobian for a multiple measurements tape.
+
+    Args:
+        tangent (tensor_like, list): tangent vector
+        jac (tensor_like, tuple): Jacobian matrix
+    Returns:
+        tensor_like: the Jacobian vector product
+
+    **Examples**
+
+    1. For a single parameter and multiple measurement (one without shape and one with shape, e.g. expval and probs):
+
+    .. code-block:: pycon
+        >>> tangent = np.array([2.0])
+        >>> jac = tuple([np.array([0.3]), np.array([0.2, 0.5])])
+        >>> qml.gradients.compute_jvp_multi(tangent, jac)
+        (np.array([0.6]), np.array([0.4, 1. ]))
+
+    2. For multiple parameters (in this case 2 parameters) and multiple measurement (one without shape and one with
+    shape, e.g. expval and probs):
+
+    .. code-block:: pycon
+        >>> tangent = np.array([1.0, 2.0])
+        >>> jac = tuple([tuple([np.array([0.3]), np.array([0.4])]), tuple([np.array([0.2, 0.5]), np.array([0.3, 0.8])]),])
+        >>> qml.gradients.compute_jvp_multi(tangent, jac)
+        (np.array([1.1]), np.array([0.8, 2.1]))
+    """
+    if jac is None:
+        return None
+    res = []
+
+    for j in jac:
+        res.append(compute_jvp_single(tangent, j))
+
+    return tuple(res)
+
+
+def jvp(tape, tangent, gradient_fn, gradient_kwargs=None):
+    r"""Generate the gradient tapes and processing function required to compute
+    the Jacobian vector products of a tape.
+    Args:
+        tape (.QuantumTape): quantum tape to differentiate
+        tangent (tensor_like, list): Gradient-output vector. Must have shape
+            matching the number of trainable parameters.
+        gradient_fn (callable): the gradient transform to use to differentiate
+            the tape
+        gradient_kwargs (dict): dictionary of keyword arguments to pass when
+            determining the gradients of tapes
+    Returns:
+        tensor_like or tuple or None: Jacobian vector product. Returns None if the tape
+        has no trainable parameters.
+    **Example**
+    #TODO: add examples
+    """
+    gradient_kwargs = gradient_kwargs or {}
+    num_params = len(tape.trainable_params)
+    num_measurements = len(tape.measurements)
+
+    if num_params == 0:
+        # The tape has no trainable parameters; the JVP
+        # is simply none.
+        return [], lambda _, num=None: None
+
+    multi_m = num_measurements > 1
+    multi_p = num_params > 1
+
+    try:
+        if math.allclose(qml.math.stack(tangent), 0):
+            # If the tangent vector is zero, then the
+            # corresponding element of the JVP will be zero,
+            # and we can avoid a quantum computation.
+
+            def func(_):  # pylint: disable=unused-argument
+                if not multi_m:
+                    res = math.convert_like(np.zeros([num_params]), tangent)
+                    res = math.cast(res, tangent.dtype)
+                    if multi_p:
+                        res = tuple(res)
```

Comment:
Can you explain this? I'm not clear on why `res` doesn't depend on the measurement shape but depends on the number of parameters.

Reply:
Very good catch, I updated it !